packages/medusa-payment-paypal/src/services/paypal-provider.ts

```diff
@@ -0,0 +1,386 @@
+                humanizeAmount(amount, currency_code),
+                currency_code
+              ),
+            },
+          },
+        ],
+      })
+
+      session_data = await this.paypal_.execute(request)
+    } catch (e) {
+      return this.buildError("An error occurred in initiatePayment", e)
+    }
+
+    return {
+      session_data,
+    }
+  }
+
+  async authorizePayment(
+    paymentSessionData: Record<string, unknown>,
+    context: Record<string, unknown>
+  ): Promise<
+    | PaymentProcessorError
+    | {
+        status: PaymentSessionStatus
+        data: PaymentProcessorSessionResponse["session_data"]
+      }
+  > {
+    try {
+      const stat = await this.getPaymentStatus(paymentSessionData)
+      return { data: paymentSessionData, status: stat }
+    } catch (error) {
+      return this.buildError("An error occurred in authorizePayment", error)
+    }
+  }
+
+  async cancelPayment(
+    paymentSessionData: Record<string, unknown>
+  ): Promise<
+    PaymentProcessorError | PaymentProcessorSessionResponse["session_data"]
+  > {
+    const order = (await this.retrievePayment(
+      paymentSessionData
+    )) as PaypalOrder
+
+    const isAlreadyCanceled = order.status === PaypalOrderStatus.VOIDED
+    const isCanceledAndFullyRefund =
+      order.status === PaypalOrderStatus.COMPLETED && !!order.invoice_id
+
+    if (isAlreadyCanceled || isCanceledAndFullyRefund) {
+      return order
+    }
+
+    try {
+      const { purchase_units } = paymentSessionData as {
+        purchase_units: PurchaseUnits
+      }
+      const isAlreadyCaptured = purchase_units.some(
+        (pu) => pu.payments.captures?.length
+      )
+
+      if (isAlreadyCaptured) {
+        const payments = purchase_units[0].payments
+
+        const payId = payments.captures[0].id
+        const request = new PayPal.payments.CapturesRefundRequest(payId)
+        await this.paypal_.execute(request)
+      } else {
+        const id = purchase_units[0].payments.authorizations[0].id
+        const request = new PayPal.payments.AuthorizationsVoidRequest(id)
+        await this.paypal_.execute(request)
+      }
+
+      return (await this.retrievePayment(
+        paymentSessionData
+      )) as unknown as PaymentProcessorSessionResponse["session_data"]
+    } catch (error) {
+      return this.buildError("An error occurred in cancelPayment", error)
+    }
+  }
+
+  async capturePayment(
+    context: PaymentProcessorContext
+  ): Promise<
+    PaymentProcessorError | PaymentProcessorSessionResponse["session_data"]
+  > {
+    const paymentSessionData = context.paymentSessionData
+    const { purchase_units } = paymentSessionData as {
+      purchase_units: PurchaseUnits
+    }
+
+    const id = purchase_units[0].payments.authorizations[0].id
+
+    try {
+      const request = new PayPal.payments.AuthorizationsCaptureRequest(id)
+      await this.paypal_.execute(request)
+      return this.retrievePayment(paymentSessionData)
```

Comment:
todo: Add missing await
```suggestion
      return (await this.retrievePayment(
        paymentSessionData
      )) as unknown as PaymentProcessorSessionResponse["session_data"]
```